app/Actions/FulfillTrade.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace App\Actions;
+
+use App\Models\Entry;
+use App\Models\Trade;
+use App\Models\Item;
+use App\Models\Character;
+use App\Models\User;
+use App\Exceptions\TradeException;
+use Illuminate\Support\Facades\Auth;
+use Lorisleiva\Actions\Concerns\AsAction;
+
+class FulfillTrade
+{
+    use AsAction;
+
+    /**
+     * Expect an array of item details, create each item and attach them to the appropriate relation objects
+     *
+     * @param Entry $entry
+     * @param array $items
+     */
+    public function handle(Trade $trade, Item $offeredItem)
+    {
+        $tradeCharacter = $trade->character;
+        $tradeItem = $trade->item;
+        $offerCharacter = $offeredItem->character;
+
+        $currentUser = Auth::user();
+        $offerUser = $offerCharacter->user;
+
+        if (!$currentUser->characters->contains($tradeCharacter) || $trade->status == Trade::STATUS_CLOSED) {
+            throw new TradeException("The trade could not be fulfilled.");
+        }
```

Comment:
This is good, but these two errors represent 2 different status codes, the first clause is one of permission, while the other is of timing. Custom exception classes can be pretty specific, usually. Something like this might be good: 
```suggestion
        if (!$currentUser->characters->contains($tradeCharacter)) {
            abort(403);
        }
        
        if ($trade->status == Trade::STATUS_CLOSED)) {
            throw new TradeClosedException();
        }
```
then the `TradeClosedException` class would specify the message and status code (400) in the construct method, which you'd override from the parent. you could also, instead of calling `abort(403)`, make a second exception class, something like `TradeNotAllowedException` that would throw a 403 for any dis-allowed trade actions, again setting the message and code in the construct.

Reply:
ah good idea, ill look into this

Reply:
done, please check that its okay